File: system-design/oop-patterns/04-factory-method-pattern/full/java/Main.java
import zoo.AnimalFactory;
import zoo.RandomAnimalFactory;
import zoo.BalancedAnimalFactory;

public class Main {
    static void populateZoo(AnimalFactory factory, int count) {
        for (int i = 0; i < count; i++) {
            System.out.println(factory.createAnimal().name());
        }
    }

    public static void main(String[] args) {
        System.out.println("Random:");
        populateZoo(new RandomAnimalFactory(), 6);

        System.out.println("Balanced:");
        populateZoo(new BalancedAnimalFactory(), 6);
    }
}

File: system-design/oop-patterns/04-factory-method-pattern/full/ts/main.ts
import type { AnimalFactory } from './zoo/AnimalFactory';
import { RandomAnimalFactory } from './zoo/RandomAnimalFactory';
import { BalancedAnimalFactory } from './zoo/BalancedAnimalFactory';

function populateZoo(factory: AnimalFactory, count: number): void {
  for (let i = 0; i < count; i++) {
    console.log(factory.createAnimal().name());
  }
}

console.log('Random:');
populateZoo(new RandomAnimalFactory(), 6);

console.log('Balanced:');
populateZoo(new BalancedAnimalFactory(), 6);
